core/mastodon/src/main/java/br/com/orcinus/orca/core/mastodon/notification/MastodonNotificationService.kt

```diff
@@ -0,0 +1,350 @@
+ */
+internal class MastodonNotificationService(
+  private val requester: Requester,
+  private val authenticationLock: SomeAuthenticationLock
+) : FirebaseMessagingService() {
+  /**
+   * IDs of [Notification]s that have been sent and have not yet been cleared by this
+   * [MastodonNotificationService].
+   *
+   * @see StatusBarNotification.getId
+   */
+  private val sentNotificationIds = mutableSetOf<Int>()
+
+  /** Private randomly generated key that identifies this [MastodonNotificationService]. */
+  private val authKey by lazy { ByteArray(16).apply(SecureRandom()::nextBytes).decodeToString() }
+
+  /**
+   * Base64-encoded [String] consisting of three sequences of bytes: an
+   * [UNCOMPRESSED_ELLIPTIC_CURVE_KEY_MARKER] and the x and y affine coordinates of the generated
+   * elliptic curve key.
+   */
+  private val publicKey by lazy {
+    KeyPairGenerator.getInstance("EC")
+      .apply { initialize(/* keysize = */ 256) }
+      .generateKeyPair()
+      .let { it.public as ECPublicKey }
+      .let { it.w.affineX.toByteArray() to it.w.affineY.toByteArray() }
+      .let { (x, y) -> byteArrayOf(UNCOMPRESSED_ELLIPTIC_CURVE_KEY_MARKER, *x, *y) }
+      .let(Base64.getUrlEncoder().withoutPadding()::encodeToString)
+  }
+
+  /**
+   * [CoroutineScope] in which remotely received payloads are converted into [MastodonNotification]
+   * DTOs and requests to subscribe to the receipt of notifications from the Mastodon server are
+   * sent.
+   *
+   * @see sendNotificationIfNotEmpty
+   * @see pushSubscription
+   */
+  @VisibleForTesting
+  var coroutineScope = CoroutineScope(Dispatchers.Default)
+    private set
+
+  /**
+   * Current [Lifecycle.State] in which this [MastodonNotificationService] is, constrained to
+   * representing only _initialized_, _created_ or _destroyed_ states. As of now, it has no actual
+   * use in production and exists only for testing purposes.
+   *
+   * @see Lifecycle.State.INITIALIZED
+   * @see Lifecycle.State.CREATED
+   * @see Lifecycle.State.DESTROYED
+   */
+  @VisibleForTesting
+  var lifecycleState = Lifecycle.State.INITIALIZED
+    private set
+
+  /**
+   * [NotificationManager] through which updates received from the server will be forwarded in the
+   * form of system notifications.
+   */
+  @VisibleForTesting
+  val notificationManager
+    get() = getSystemService<NotificationManager>() as NotificationManager
+
+  /**
+   * Whether the Firebase SDK is currently running.
+   *
+   * @see startFirebaseSdk
+   * @see stopFirebaseSdkIfRunning
+   */
+  @VisibleForTesting
+  val isFirebaseSdkRunning
+    get() =
+      try {
+        Firebase.app
+        true
+      } catch (_: IllegalStateException) {
+        false
+      }
+
+  @Throws(Module.DependencyNotInjectedException::class)
+  constructor() :
+    this(requester = Injector.get(), Injector.from<CoreModule>().authenticationLock()) {
+    setCoroutineContext(SupervisorJob() + Dispatchers.IO)
```

Comment:
An [`Injector.ModuleNotRegisteredException`](https://github.com/orcinusbr/orca-android/blob/7cca5d4e7050e975e52eb73eee18227be19e7516/std/injector/src/main/java/br/com/orcinus/orca/std/injector/Injector.kt#L48) can also be thrown by this constructor if a [`CoreModule`](https://github.com/orcinusbr/orca-android/blob/7cca5d4e7050e975e52eb73eee18227be19e7516/core-module/src/main/java/br/com/orcinus/orca/core/module/CoreModule.kt) is not registered.